File: java/124_BinaryTreeMaximumPathSum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private int res = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        maxRootSum(root);
        return res;
    }
    
    public int maxRootSum(TreeNode root) {
	if(root == null) {
	    return 0;
	}
	int curVal = root.val;
	int leftMaxSum = maxRootSum(root.left);
	int rightMaxSum = maxRootSum(root.right);
		
	if(leftMaxSum > 0) {
	    curVal += leftMaxSum;
	}
	if(rightMaxSum > 0) {
	    curVal += rightMaxSum;
	}
	if(curVal > res) {
	    res = curVal;
	}
        // return res; !错误理解题目，wa了很多次
	return Math.max(root.val, Math.max(root.val + leftMaxSum, root.val + rightMaxSum));
    }
}

File: c++/124_BinaryTreeMaximumPathSum.cpp
class Solution {
public:
    int res;
    int maxPathSum(TreeNode* root) {
        res = INT_MIN;
        dfs(root);
        return res;
    }
    
    int dfs(TreeNode* u) {
        if (!u) return 0;
        int left = max(0, dfs(u->left));
        int right = max(0, dfs(u->right));
        res = max(res, u->val + left + right);
        return u->val + max(left, right);
    }
};
